
Hello! This is a tutorial on how to use the package easydars, which is meant to conduct downstream analysis of bulk RNA sequencing data, including heatmaps, volcano plots, and preparing gmt (gene matrix transposed) files and rankings for fast gene set enrichment analysis (fgsea). 

First, run this chunk to install the package. This also downloads the test data files I use in this tutorial. 

```{r}
library(devtools)
devtools::install_github("eitm-org/easydars")
library(easydars)
```

Yay! Now we can load in other necessary packages. 

```{r}
library(DESeq2)
library(EnhancedVolcano)
library(dplyr)
library(pheatmap)
library(janitor)
library(tidyverse)
library(purrr)
library(fgsea)
library(org.Mm.eg.db) # genome wide annotations for mouse - replace if using human data (needed for gene ID to name mapping)
```

Next, let's load in our data. The first form of data we need is counts data, which has the raw counts of each gene. In this example, the first column contains gene IDs, and the other columns are samples. These sample columns contain the raw RNA counts for each gene. We can use clean_names to make sure names in the dataframe have unique names and only contain characters, numbers, and letters. 
```{r}
count_df <- read.table(system.file("extdata","Condition_genotype-WT-KO-batch.normalised_counts.tsv", package = "easydars"), header = TRUE) %>%
  clean_names()
```

This dataset only contains gene IDs, but we are interested in gene names as well. We can use the mouse genome wide annotations to map our existing gene IDs to their corresponding gene names (keeping the gene ID if there is no matching name). This step can be skipped if counts data already includes gene_id and gene_name columns.
```{r}
ens <- count_df$gene_id
symbols <- mapIds(org.Mm.eg.db, keys = ens,
                  column = 'SYMBOL', keytype = 'ENSEMBL')
symbols <- symbols[!is.na(symbols)]
symbols <- symbols[match(count_df$gene_id, names(symbols))]
count_df$gene_name <- symbols
count_df$gene_name <- ifelse(is.na(count_df$gene_name), count_df$gene_id, count_df$gene_name)
```

The next data we need is col data, which contains metadata about each sample. 
Here, I take my sample identifier, the column sample_names, and turn it into the rownames of the column. 
```{r}
col_data <- read.csv(system.file("extdata","Condition_genotype-WT-KO-metadata.csv", package = "easydars")) %>%
  clean_names()
rownames(col_data) <- col_data$sample_names
col_data <- col_data %>%
  dplyr::select(-sample_names)
```

We must make sure that the number of columns of our counts data is equal to the number of rows in our col data, a.k.a. we have the same number of samples in both. Right now, our count_df has two extra columns: gene_id and gene_name. We can fix that by making gene_id into row names, and dropping gene_name. Keep the original count_df, as it will be useful later when we need to make plots with our gene names. 
```{r}
count_mat <- count_df %>% 
  column_to_rownames(var = "gene_id") %>%
  dplyr::select(-gene_name) %>%
  clean_names() 
```

Next, we can run differential expression analysis using DESeq2. First, we make the dataset with DESeqDataSetFromMatrix(). countData is the dataframe of our raw counts with the modified columns (count_mat), colData is our metadata (col_data), and design will be ~ followed by the column of col_data we are interested in (in this case, I chose sample_genotype). Then, we can run that dataset in DESeq(). 
```{r}
genotype_dds <- DESeqDataSetFromMatrix(countData = round(count_mat),
                              colData = col_data,
                              design = ~ sample_genotype)
genotype_DEs <- DESeq(genotype_dds)
```

Great! Now that we've run DESeq2, we can use resultsNames(), which will output the estimated effects (coefficients) of our design. 
```{r}
resultsNames(genotype_DEs)
```
resultsNames() outputted "sample_genotype_WT_vs_KO" - this is the estimated effects of our design, examining differential expression between two genotypes, WT and KO. 

Next, lets extract a results table from our DESeq analysis, which will contain base means, log2fold changes, p-vals, adjusted p-vals, among other things. Input the DESeq dataset (ie. genotype_DEs) and your coefficient (ie. name = "sample_genotype_WT_vs_KO"). You can specify a number of parameters here. For the method for adjusting p-values, we will use the Benjamini-Hochberg procedure (pAdjustMethod = "BH"), a significance cutoff (alpha) of 0.05, and independent filtering. 
```{r}
res <- results(genotype_DEs,
               pAdjustMethod = "BH",  
               name = "sample_genotype_WT_vs_KO",  
               alpha = 0.05,          
               independentFiltering = TRUE)
```

Now we have all the inputs we need to use easydars! 

Let's start by making a Principal Component Analysis (PCA) plot to visualize our groups. The inputs you need are:
- the DESeq dataset (ie. org_DEs)
- the name of your coefficient/design of experiment (ie. "sample_genotype")
- the number of top features to use (default 500)
- the PC to be on x-axis (default 1)
- the PC to be on y-axis (default 2)
```{r}
genotype_PCA <- DE_pca(dds = genotype_DEs, 
                  xvar = "sample_genotype")
print(genotype_PCA)
```

Now, let's make a volcano plot, which has the log2fold change on the x axis, and the -log(p-adjusted value) on the y axis. The inputs you need are:
- the counts data that contains both a gene_name and gene_id column (ie. count_df)
- the DESeq dataset (ie. genotype_DEs)
- the results table that we just made above (ie. res)
- the shrinkage method (apeglm and ashr are recommended)
- the coefficient or design of your experiment (ie. "sample_genotype")
- the name of your comparison (ie."sample_genotype_WT_vs_KO")
- the names of the groups you are comparing (ie. "WT" and "KO")
```{r, fig.height = 8, fig.width = 15}
genotype_volcanoplot <- DE_table_volcano(count_df = count_df,
                                    dds = genotype_DEs,
                                    res = res,
                                    shrinkage_method = "apeglm",
                                    xvar = "sample_genotype",
                                    name_of_comparison = "sample_genotype_WT_vs_KO",
                                    first_group = "WT",
                                    second_group = "KO"
                                                                )
print(genotype_volcanoplot)
```

You can also extract the top differentially expressed genes (those with the lowest p-adjusted values)
```{r}
DE_genes <- genotype_volcanoplot$de_genes
print(DE_genes)
```

Next, we can make a heatmap. The inputs you need are:
- the results table that we just made above (ie. res)
- the DESeq dataset (ie. genotype_DEs)
- the col data aka the sample metadata (ie. col_data)
- the counts data that contains both a gene_name and gene_id column (ie. count_df)
- the coefficient or design of your experiment (ie. "sample_genotype")
- the name of your comparison (ie."sample_genotype_WT_vs_KO")
By default, the heatmap will show the top 50 genes with the lowest p-adjusted values. You can specify how many genes you want to be displayed by adjusting the top_n parameter. In this example, I chose 10. 
```{r}
genotype_heatmap_top10 <- DE_heat_map(count_df = count_df,
                                                             col_data = col_data,
                                                             dds = genotype_DEs,
                                                             res = res,
                                                             xvar = "sample_genotype",
                                                             name_of_comparison = "sample_genotype_WT_vs_KO",
                                                             top_n = 10
                                                             )
print(genotype_heatmap_top10)
```

You can also specify genes names of interest to be displayed on the heatmap.
```{r}
genes_of_interest = c("Prex1","Lrp3","Tspan6", "Mical1")
genotype_heatmap_genes_of_interest <- DE_heat_map(count_df = count_df,
                                                             col_data = col_data,
                                                             dds = genotype_DEs,
                                                             res = res,
                                                             xvar = "sample_genotype",
                                                             name_of_comparison = "sample_genotype_WT_vs_KO",
                                                             genes_of_interest = genes_of_interest
                                                             )
print(genotype_heatmap_genes_of_interest)
```
Yay! Now, we can prepare to run fast gene set enrichment analysis (fgsea). We can use information that we previously extracted in the volcano plot function. Specifically, we need the log2fold changes, p values (not adjusted p-values, which are discrete variables, as this could lead to ranking ties), gene_names, all of which can be found in the shrunken results table automatically generate from the DE_table_volcano function. These are all the inputs you need to generate a ranked list of genes with the prepare_rankings function! 
```{r}
fc <- genotype_volcanoplot$res_shrink$log2FoldChange
p_val <- genotype_volcanoplot$res_shrink$pvalue
rownames <- genotype_volcanoplot$res_shrink$gene_name

rankings <- prepare_rankings(log2foldchange = fc, pvalues = p_val, row_names = rownames)
print(rankings)
```

With the ranked genes, we are one step closer to being able to run fgsea! Next, we need to prepare our gmt files, containing gene sets. First, we can extract a list of gene names from our counts data. Then, select the .gmt files you would like to prepare. Make sure the list of genes is uppercase if it isn't already. 
```{r}
my_genes <- toupper(count_df$gene_name)
gmt_file <- system.file("extdata","mh.all.v2022.1.Mm.symbols.gmt", package = "easydars")
```

Next, we can use the prepare_gmt function, which take in our gmt_file path (gmt_file), along with our list of gene names from count_df (my_genes). 
```{r}
pathways <- prepare_gmt(gmt_file, my_genes)
```

All that's left to do is run fgsea! The parameters of fgsea are a list of gene sets to check (pathways) and a named vector of gene-level stats with the same names as our gene sets (we just created this with our prepare_rankings function, rankings).
```{r}
genotype_fgsea_res <- fgsea(pathways, rankings)
print(genotype_fgsea_res)
```

