
Hello! This is a tutorial on how to use easydars, which is meant to conduct downstream analysis of bulk RNA sequencing data, including heatmaps, volcano plots, and preparing gmt (gene matrix transposed) files and rankings for fast gene set enrichment analysis (fgsea). 

You can follow along with this tutorial by git cloning the eitm-org/metastatic-niche-DE repository onto your desktop, and changing the paths below, so that you can load the same data too! 

First, run this chunk to install the package. 
```{r}
library(devtools)
devtools::install_github("eitm-org/easydars")
library(easydars)
```

Yay! Now we can load in other necessary packages. 

```{r}
library(DESeq2)
library(EnhancedVolcano)
library(dplyr)
library(pheatmap)
library(janitor)
library(tidyverse)
library(purrr)
library(fgsea)
```

Next, let's load in our data. The first form of data we need is counts data, which has the raw counts of each gene. In this example, the first two columns are gene id's and gene names, and the other columns are samples. These sample columns contain the raw RNA counts for each gene. We can use clean_names to make sure names in the dataframe have unique names and only contain characters, numbers, and letters. 
```{r}
count_df <- read.table(file = "/Users/eli/Desktop/metastatic-niche-DE/data/salmon.merged.gene_counts_length_scaled.tsv", header = TRUE) %>%
  clean_names()
```

The next data we need is col data, which contains metadata about each sample.
```{r}
col_data0 <- read_csv(file = "/Users/eli/Desktop/metastatic-niche-DE/data/mp_rna_seq_metadata_052224.csv",
                      show_col_types = FALSE
                      ) %>% 
  clean_names()
```

Select your column of interest from your col data, which is the variable you want to compare in your differential expression analysis. In this case, I am looking at organoids_cell_line. 
```{r}
col_data_org <- col_data0 %>%
  dplyr::select(eit_id, organoids_cell_line)
```

We must make sure that the number of columns of our counts data is equal to the number of rows in our col data. Right now, our count_df has two extra columns: gene_id and gene_name. We can fix that by making gene_id into row names, and dropping gene_name. Keep the original count_df, as it will be useful later for mapping gene IDs with their corresponding names. 
```{r}
count_mat <- count_df %>% 
  column_to_rownames(var = "gene_id") %>%
  clean_names() %>%
  dplyr::select(-gene_name)
```

Next, we can run differential expression analysis using DESeq2. First, we make the dataset with DESeqDataSetFromMatrix(). countData is the dataframe of our raw counts with the modified columns (count_mat), colData is our metadata (col_data_org), and design will be ~ followed by the column of col_data_org we are interested in (organoids_cell_line). Then, we can run that dataset in DESeq(). 
```{r}
org_dds <- DESeqDataSetFromMatrix(countData = round(count_mat),
                              colData = col_data_org,
                              design = ~ organoids_cell_line)
org_DEs <- DESeq(org_dds)
```

Great! Now that we've run DESeq2, we can use resultsNames(), which will output the estimated effects (coefficients) of our design. 
```{r}
resultsNames(org_DEs)
```

Next, lets extract a results table from our DESeq analysis, which will contain base means, log2fold changes, p-vals, adjusted p-vals, among other things. Input the DESeq dataset (ie. org_DEs) and your coefficient (ie. name = "organoids_cell_line_MT3B_vs_MM1). You can specify a number of parameters here. For the method for adjusting p-values, we will use the Benjamini-Hochberg procedure (pAdjustMethod = "BH"), a significance cutoff (alpha) of 0.05, and independent filtering. 
```{r}
res <- results(org_DEs,
               pAdjustMethod = "BH",  
               name = "organoids_cell_line_MT3B_vs_MM1",  
               alpha = 0.05,          
               independentFiltering = TRUE)
```

Now we have all the inputs we need to use easydars! 

Let's start by making a Principal Component Analysis (PCA) plot to visualize our groups. The inputs you need are:
- the DESeq dataset (ie. org_DEs)
- the name of your coefficient/design of experiment (ie. "organoids_cell_line)
- the number of top features to use (default 500)
- the PC to be on x-axis (default 1)
- the PC to be on y-axis (default 2)
```{r}
organoids_cell_line_MT3B_vs_MM1_PCA <- DE_pca(dds = org_DEs, 
                                              xvar = "organoids_cell_line"
                                              )
print(organoids_cell_line_MT3B_vs_MM1_PCA)
```

Now, let's make a volcano plot, which has the log2fold change on the x axis, and the -log(p-adjusted value) on the y axis. The inputs you need are:
- the counts data that contains both a gene_name and gene_id column (ie. count_df)
- the DESeq dataset (ie. org_DEs)
- the results table that we just made above (ie. res)
- the coefficient or design of your experiment (ie. "organoids_cell_line")
- the name of your comparison (ie."organoids_cell_line_MT3B_vs_MM1")
- the names of the groups you are comparing (ie. "MT3B" and "MM1")
You will also need to specify your shrinkage method (apeglm and ashr are recommended). 
```{r, fig. height = 20}
organoids_cell_line_MT3B_vs_MM1_volcanoplot <- DE_table_volcano(count_df = count_df,
                                                                dds = org_DEs,
                                                                res = res,
                                                                shrinkage_method = "apeglm",
                                                                xvar = "organoids_cell_line",
                                                                name_of_comparison = "organoids_cell_line_MT3B_vs_MM1",
                                                                first_group = "MT3B",
                                                                second_group = "MM1"
                                                                )
                                                         
print(organoids_cell_line_MT3B_vs_MM1_volcanoplot)
```
You can also extract the top differentially expressed genes (those with the lowest p-adjusted values)
```{r}
DE_genes <- organoids_cell_line_MT3B_vs_MM1_volcanoplot$de_genes
print(DE_genes)
```

Next, we can make a heatmap. The inputs you need are:
- the results table that we just made above (ie. res)
- the DESeq dataset (ie. org_DEs)
- the col data aka the sample metadata (ie. col_data_org)
- the counts data that contains both a gene_name and gene_id column (ie. count_df)
- the coefficient or design of your experiment (ie. "organoids_cell_line")
- the name of your comparison (ie."organoids_cell_line_MT3B_vs_MM1")
By default, the heatmap will show the top 50 genes with the lowest p-adjusted values. You can specify how many genes you want to be displayed by adjusting the top_n parameter. In this example, I chose 15. 
```{r}
organoids_cell_line_MT3B_vs_MM1_heatmap_top15 <- DE_heat_map(count_df = count_df,
                                                             col_data = col_data_org,
                                                             dds = org_DEs,
                                                             res = res,
                                                             xvar = "organoids_cell_line",
                                                             name_of_comparison = "organoids_cell_line_MT3B_vs_MM1",
                                                             top_n = 15
                                                             )
print(organoids_cell_line_MT3B_vs_MM1_heatmap_top15)
```

You can also specify genes names of interest to be displayed on the heatmap.
```{r}
genes_of_interest = c("Tnf", "Fbp2", "Dlx3", "Fer")
organoids_cell_line_MT3B_vs_MM1_heatmap_genes_of_interest <- DE_heat_map(count_df = count_df,
                                                             col_data = col_data_org,
                                                             dds = org_DEs,
                                                             res = res,
                                                             name_of_comparison = "organoids_cell_line_MT3B_vs_MM1",
                                                             genes_of_interest = genes_of_interest
                                                             )
print(organoids_cell_line_MT3B_vs_MM1_heatmap_genes_of_interest)
```
Yay! Now, we can prepare to run fast gene set enrichment analysis (fgsea). We can use information that we previously extracted in the volcano plot function. Specifically, we need the log2fold changes, p values (not adjusted p-values, which are discrete variables, as this could lead to ranking ties), gene_names, all of which can be found in the shrunken results table automatically generate from the DE_table_volcano function. These are all the inputs you need to generate a ranked list of genes with the prepare_rankings function! 
```{r}
fc <- organoids_cell_line_MT3B_vs_MM1_volcanoplot$res_shrink$log2FoldChange
p_val <- organoids_cell_line_MT3B_vs_MM1_volcanoplot$res_shrink$pvalue
rownames <- organoids_cell_line_MT3B_vs_MM1_volcanoplot$res_shrink$gene_name

rankings <- prepare_rankings(log2foldchange = fc, pvalues = p_val, row_names = rownames)
print(rankings)
```

With the ranked genes, we are one step closer to being able to run fgsea! Next, we need to prepare our gmt files,containing gene sets.
First, we can extract a list of gene names from our counts data. Then, list out all the .gmt files you would like to prepare. 
```{r}
my_genes <- count_df$gene_name
gmt_files <- list.files("/Users/eli/Desktop/metastatic-niche-DE/data/cell_type", pattern = ".gmt", full.names = TRUE)
```

Next, we can use the map() function to apply the prepare_gmt function to all the gmt files we have. 
```{r}
all_gmt_files <- map(gmt_files, prepare_gmt, my_genes, savefile = TRUE)
```

All that's left to do is run fgsea! Using the map() function again, we can run fgsea() on every gmt file that we prepared! The other parameters for fgsea() are stats, which are the rankings we created using the prepare_rankings function above, 
```{r}
organoids_cell_line_MT3B_vs_MM1_fgsea_res <- map(
               all_gmt_files,
               fgsea,
               stats = rankings,
)
print(organoids_cell_line_MT3B_vs_MM1_fgsea_res)
```

